src/Platform/Microsoft.Testing.Extensions.AzureDevOps/AzureDevOpsReporter.cs

```diff
@@ -0,0 +1,177 @@
+    private string _severity = "error";
+
+    public AzureDevOpsReporter(
+        ICommandLineOptions commandLine,
+        IEnvironment environment,
+        IOutputDevice outputDisplay)
+    {
+        _commandLine = commandLine;
+        _environment = environment;
+        _outputDisplay = outputDisplay;
+    }
+
+    public Type[] DataTypesConsumed { get; } =
+    [
+        typeof(TestNodeUpdateMessage)
+    ];
+
+    public Type[] DataTypesProduced { get; } = [typeof(SessionFileArtifact)];
+
+    /// <inheritdoc />
+    public string Uid { get; } = nameof(AzureDevOpsReporter);
+
+    /// <inheritdoc />
+    public string Version { get; } = AppVersion.DefaultSemVer;
+
+    /// <inheritdoc />
+    public string DisplayName { get; } = AzureDevOpsResources.DisplayName;
+
+    /// <inheritdoc />
+    public string Description { get; } = AzureDevOpsResources.Description;
+
+    /// <inheritdoc />
+    public Task<bool> IsEnabledAsync()
+    {
+        bool isEnabled = _commandLine.IsOptionSet(AzureDevOpsCommandLineOptions.AzureDevOpsOptionName)
+            && string.Equals(_environment.GetEnvironmentVariable("TF_BUILD"), "true", StringComparison.OrdinalIgnoreCase);
+
+        if (isEnabled)
+        {
+            bool found = _commandLine.TryGetOptionArgumentList(AzureDevOpsCommandLineOptions.AzureDevOpsReportSeverity, out string[]? arguments);
+            if (found && arguments?.Length > 0)
+            {
+                if (string.Equals(arguments[0], "warning", StringComparison.OrdinalIgnoreCase))
+                {
+                    _severity = "warning";
+                }
+            }
+        }
+
+        return Task.FromResult(isEnabled);
+    }
+
+    public async Task ConsumeAsync(IDataProducer dataProducer, IData value, CancellationToken cancellationToken)
+    {
+        if (cancellationToken.IsCancellationRequested)
+        {
+            return;
+        }
+
+        try
+        {
+            if (value is not TestNodeUpdateMessage nodeUpdateMessage)
+            {
+                return;
+            }
+
+            TestNodeStateProperty nodeState = nodeUpdateMessage.TestNode.Properties.Single<TestNodeStateProperty>();
+
+            switch (nodeState)
+            {
+                case FailedTestNodeStateProperty failed:
+                    await WriteExceptionAsync(failed.Explanation, failed.Exception);
+                    break;
+                case ErrorTestNodeStateProperty error:
+                    await WriteExceptionAsync(error.Explanation, error.Exception);
+                    break;
+                case CancelledTestNodeStateProperty cancelled:
+                    await WriteExceptionAsync(cancelled.Explanation, cancelled.Exception);
+                    break;
+                case TimeoutTestNodeStateProperty timeout:
+                    await WriteExceptionAsync(timeout.Explanation, timeout.Exception);
+                    break;
+            }
```

Comment:
I think we may be doing this somewhere else. Maybe time to extract an internal helper?

Reply:
There already is TestNodePropertiesCategories.WellKnownTestNodeTestRunOutcomeFailedProperties but that cannot touch the node state, and it is not practical to implement the helper in its terms, because once we have a switch, it is extra work to check if the TestNodePropertiesCategories.WellKnownTestNodeTestRunOutcomeFailedProperties array has the state value.

And I would like those two to not become out of sync.

Moving to new issue.